{"$message_type":"diagnostic","message":"unresolved import `crate::benchmark`","code":{"code":"E0432","explanation":"An import was unresolved.\n\nErroneous code example:\n\n```compile_fail,E0432\nuse something::Foo; // error: unresolved import `something::Foo`.\n```\n\nIn Rust 2015, paths in `use` statements are relative to the crate root. To\nimport items relative to the current and parent modules, use the `self::` and\n`super::` prefixes, respectively.\n\nIn Rust 2018 or later, paths in `use` statements are relative to the current\nmodule unless they begin with the name of a crate or a literal `crate::`, in\nwhich case they start from the crate root. As in Rust 2015 code, the `self::`\nand `super::` prefixes refer to the current and parent modules respectively.\n\nAlso verify that you didn't misspell the import name and that the import exists\nin the module from where you tried to import it. Example:\n\n```\nuse self::something::Foo; // Ok.\n\nmod something {\n    pub struct Foo;\n}\n# fn main() {}\n```\n\nIf you tried to use a module from an external crate and are using Rust 2015,\nyou may have missed the `extern crate` declaration (which is usually placed in\nthe crate root):\n\n```edition2015\nextern crate core; // Required to use the `core` crate in Rust 2015.\n\nuse core::any;\n# fn main() {}\n```\n\nSince Rust 2018 the `extern crate` declaration is not required and\nyou can instead just `use` it:\n\n```edition2018\nuse core::any; // No extern crate required in Rust 2018.\n# fn main() {}\n```\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":107,"byte_end":116,"line_start":4,"line_end":4,"column_start":12,"column_end":21,"is_primary":true,"text":[{"text":"use crate::benchmark::DummyArrayVecBenchmark;","highlight_start":12,"highlight_end":21}],"label":"could not find `benchmark` in the crate root","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0432]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: unresolved import `crate::benchmark`\u001b[0m\n\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:4:12\u001b[0m\n\u001b[0m  \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m4\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0muse crate::benchmark::DummyArrayVecBenchmark;\u001b[0m\n\u001b[0m  \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m            \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mcould not find `benchmark` in the crate root\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"unused import: `wrap_pymodule`","code":{"code":"unused_imports","explanation":null},"level":"warning","spans":[{"file_name":"src\\pylib.rs","byte_start":51,"byte_end":64,"line_start":2,"line_end":2,"column_start":29,"column_end":42,"is_primary":true,"text":[{"text":"use pyo3::{wrap_pyfunction, wrap_pymodule};","highlight_start":29,"highlight_end":42}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"`#[warn(unused_imports)]` on by default","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"remove the unused import","code":null,"level":"help","spans":[{"file_name":"src\\pylib.rs","byte_start":33,"byte_end":34,"line_start":2,"line_end":2,"column_start":11,"column_end":12,"is_primary":true,"text":[{"text":"use pyo3::{wrap_pyfunction, wrap_pymodule};","highlight_start":11,"highlight_end":12}],"label":null,"suggested_replacement":"","suggestion_applicability":"MachineApplicable","expansion":null},{"file_name":"src\\pylib.rs","byte_start":49,"byte_end":64,"line_start":2,"line_end":2,"column_start":27,"column_end":42,"is_primary":true,"text":[{"text":"use pyo3::{wrap_pyfunction, wrap_pymodule};","highlight_start":27,"highlight_end":42}],"label":null,"suggested_replacement":"","suggestion_applicability":"MachineApplicable","expansion":null},{"file_name":"src\\pylib.rs","byte_start":64,"byte_end":65,"line_start":2,"line_end":2,"column_start":42,"column_end":43,"is_primary":true,"text":[{"text":"use pyo3::{wrap_pyfunction, wrap_pymodule};","highlight_start":42,"highlight_end":43}],"label":null,"suggested_replacement":"","suggestion_applicability":"MachineApplicable","expansion":null}],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;11mwarning\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: unused import: `wrap_pymodule`\u001b[0m\n\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:2:29\u001b[0m\n\u001b[0m  \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m2\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0muse pyo3::{wrap_pyfunction, wrap_pymodule};\u001b[0m\n\u001b[0m  \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                             \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;11m^^^^^^^^^^^^^\u001b[0m\n\u001b[0m  \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m  \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: `#[warn(unused_imports)]` on by default\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope","code":{"code":"E0599","explanation":"This error occurs when a method is used on a type which doesn't implement it:\n\nErroneous code example:\n\n```compile_fail,E0599\nstruct Mouth;\n\nlet x = Mouth;\nx.chocolate(); // error: no method named `chocolate` found for type `Mouth`\n               //        in the current scope\n```\n\nIn this case, you need to implement the `chocolate` method to fix the error:\n\n```\nstruct Mouth;\n\nimpl Mouth {\n    fn chocolate(&self) { // We implement the `chocolate` method here.\n        println!(\"Hmmm! I love chocolate!\");\n    }\n}\n\nlet x = Mouth;\nx.chocolate(); // ok!\n```\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":3648,"byte_end":3660,"line_start":111,"line_end":111,"column_start":7,"column_end":19,"is_primary":true,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_initialize, m)?)?;","highlight_start":7,"highlight_end":19}],"label":"method not found in `&PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0599]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:111:7\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m111\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_initialize, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m       \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mmethod not found in `&PyModule`\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied","code":{"code":"E0277","explanation":"You tried to use a type which doesn't implement some trait in a place which\nexpected that trait.\n\nErroneous code example:\n\n```compile_fail,E0277\n// here we declare the Foo trait with a bar method\ntrait Foo {\n    fn bar(&self);\n}\n\n// we now declare a function which takes an object implementing the Foo trait\nfn some_func<T: Foo>(foo: T) {\n    foo.bar();\n}\n\nfn main() {\n    // we now call the method with the i32 type, which doesn't implement\n    // the Foo trait\n    some_func(5i32); // error: the trait bound `i32 : Foo` is not satisfied\n}\n```\n\nIn order to fix this error, verify that the type you're using does implement\nthe trait. Example:\n\n```\ntrait Foo {\n    fn bar(&self);\n}\n\n// we implement the trait on the i32 type\nimpl Foo for i32 {\n    fn bar(&self) {}\n}\n\nfn some_func<T: Foo>(foo: T) {\n    foo.bar(); // we can now use this method since i32 implements the\n               // Foo trait\n}\n\nfn main() {\n    some_func(5i32); // ok!\n}\n```\n\nOr in a generic context, an erroneous code example would look like:\n\n```compile_fail,E0277\nfn some_func<T>(foo: T) {\n    println!(\"{:?}\", foo); // error: the trait `core::fmt::Debug` is not\n                           //        implemented for the type `T`\n}\n\nfn main() {\n    // We now call the method with the i32 type,\n    // which *does* implement the Debug trait.\n    some_func(5i32);\n}\n```\n\nNote that the error here is in the definition of the generic function. Although\nwe only call it with a parameter that does implement `Debug`, the compiler\nstill rejects the function. It must work with all possible input types. In\norder to make this example compile, we need to restrict the generic type we're\naccepting:\n\n```\nuse std::fmt;\n\n// Restrict the input type to types that implement Debug.\nfn some_func<T: fmt::Debug>(foo: T) {\n    println!(\"{:?}\", foo);\n}\n\nfn main() {\n    // Calling the method is still fine, as i32 implements Debug.\n    some_func(5i32);\n\n    // This would fail to compile now:\n    // struct WithoutDebug;\n    // some_func(WithoutDebug);\n}\n```\n\nRust only looks at the signature of the called function, as such it must\nalready specify all requirements that will be used for every type parameter.\n"},"level":"error","spans":[{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5200,"byte_end":5212,"line_start":153,"line_end":153,"column_start":13,"column_end":25,"is_primary":true,"text":[{"text":"            py_or_module,","highlight_start":13,"highlight_end":25}],"label":"the trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3661,"byte_end":3702,"line_start":111,"line_end":111,"column_start":20,"column_end":61,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_initialize, m)?)?;","highlight_start":20,"highlight_end":61}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}},{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5125,"byte_end":5186,"line_start":152,"line_end":152,"column_start":9,"column_end":70,"is_primary":false,"text":[{"text":"        $crate::impl_::pyfunction::WrapPyFunctionArg::wrap_pyfunction(","highlight_start":9,"highlight_end":70}],"label":"required by a bound introduced by this call","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3661,"byte_end":3702,"line_start":111,"line_end":111,"column_start":20,"column_end":61,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_initialize, m)?)?;","highlight_start":20,"highlight_end":61}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}}],"children":[{"message":"the following other types implement trait `WrapPyFunctionArg<'py, T>`:\n  &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  &pyo3::Bound<'py, pyo3::types::PyModule>\n  OnlyBound<T>\n  pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  pyo3::Bound<'py, pyo3::types::PyModule>\n  pyo3::Python<'py>","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0277]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:111:20\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m111\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_initialize, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mthe trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mrequired by a bound introduced by this call\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mhelp\u001b[0m\u001b[0m: the following other types implement trait `WrapPyFunctionArg<'py, T>`:\u001b[0m\n\u001b[0m              &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              &pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              OnlyBound<T>\u001b[0m\n\u001b[0m              pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Python<'py>\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: this error originates in the macro `wrap_pyfunction` (in Nightly builds, run with -Z macro-backtrace for more info)\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope","code":{"code":"E0599","explanation":"This error occurs when a method is used on a type which doesn't implement it:\n\nErroneous code example:\n\n```compile_fail,E0599\nstruct Mouth;\n\nlet x = Mouth;\nx.chocolate(); // error: no method named `chocolate` found for type `Mouth`\n               //        in the current scope\n```\n\nIn this case, you need to implement the `chocolate` method to fix the error:\n\n```\nstruct Mouth;\n\nimpl Mouth {\n    fn chocolate(&self) { // We implement the `chocolate` method here.\n        println!(\"Hmmm! I love chocolate!\");\n    }\n}\n\nlet x = Mouth;\nx.chocolate(); // ok!\n```\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":3714,"byte_end":3726,"line_start":112,"line_end":112,"column_start":7,"column_end":19,"is_primary":true,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_add, m)?)?;","highlight_start":7,"highlight_end":19}],"label":"method not found in `&PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0599]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:112:7\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m112\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_add, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m       \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mmethod not found in `&PyModule`\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied","code":{"code":"E0277","explanation":"You tried to use a type which doesn't implement some trait in a place which\nexpected that trait.\n\nErroneous code example:\n\n```compile_fail,E0277\n// here we declare the Foo trait with a bar method\ntrait Foo {\n    fn bar(&self);\n}\n\n// we now declare a function which takes an object implementing the Foo trait\nfn some_func<T: Foo>(foo: T) {\n    foo.bar();\n}\n\nfn main() {\n    // we now call the method with the i32 type, which doesn't implement\n    // the Foo trait\n    some_func(5i32); // error: the trait bound `i32 : Foo` is not satisfied\n}\n```\n\nIn order to fix this error, verify that the type you're using does implement\nthe trait. Example:\n\n```\ntrait Foo {\n    fn bar(&self);\n}\n\n// we implement the trait on the i32 type\nimpl Foo for i32 {\n    fn bar(&self) {}\n}\n\nfn some_func<T: Foo>(foo: T) {\n    foo.bar(); // we can now use this method since i32 implements the\n               // Foo trait\n}\n\nfn main() {\n    some_func(5i32); // ok!\n}\n```\n\nOr in a generic context, an erroneous code example would look like:\n\n```compile_fail,E0277\nfn some_func<T>(foo: T) {\n    println!(\"{:?}\", foo); // error: the trait `core::fmt::Debug` is not\n                           //        implemented for the type `T`\n}\n\nfn main() {\n    // We now call the method with the i32 type,\n    // which *does* implement the Debug trait.\n    some_func(5i32);\n}\n```\n\nNote that the error here is in the definition of the generic function. Although\nwe only call it with a parameter that does implement `Debug`, the compiler\nstill rejects the function. It must work with all possible input types. In\norder to make this example compile, we need to restrict the generic type we're\naccepting:\n\n```\nuse std::fmt;\n\n// Restrict the input type to types that implement Debug.\nfn some_func<T: fmt::Debug>(foo: T) {\n    println!(\"{:?}\", foo);\n}\n\nfn main() {\n    // Calling the method is still fine, as i32 implements Debug.\n    some_func(5i32);\n\n    // This would fail to compile now:\n    // struct WithoutDebug;\n    // some_func(WithoutDebug);\n}\n```\n\nRust only looks at the signature of the called function, as such it must\nalready specify all requirements that will be used for every type parameter.\n"},"level":"error","spans":[{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5200,"byte_end":5212,"line_start":153,"line_end":153,"column_start":13,"column_end":25,"is_primary":true,"text":[{"text":"            py_or_module,","highlight_start":13,"highlight_end":25}],"label":"the trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3727,"byte_end":3761,"line_start":112,"line_end":112,"column_start":20,"column_end":54,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_add, m)?)?;","highlight_start":20,"highlight_end":54}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}},{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5125,"byte_end":5186,"line_start":152,"line_end":152,"column_start":9,"column_end":70,"is_primary":false,"text":[{"text":"        $crate::impl_::pyfunction::WrapPyFunctionArg::wrap_pyfunction(","highlight_start":9,"highlight_end":70}],"label":"required by a bound introduced by this call","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3727,"byte_end":3761,"line_start":112,"line_end":112,"column_start":20,"column_end":54,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_add, m)?)?;","highlight_start":20,"highlight_end":54}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}}],"children":[{"message":"the following other types implement trait `WrapPyFunctionArg<'py, T>`:\n  &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  &pyo3::Bound<'py, pyo3::types::PyModule>\n  OnlyBound<T>\n  pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  pyo3::Bound<'py, pyo3::types::PyModule>\n  pyo3::Python<'py>","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0277]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:112:20\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m112\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_add, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mthe trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mrequired by a bound introduced by this call\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mhelp\u001b[0m\u001b[0m: the following other types implement trait `WrapPyFunctionArg<'py, T>`:\u001b[0m\n\u001b[0m              &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              &pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              OnlyBound<T>\u001b[0m\n\u001b[0m              pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Python<'py>\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: this error originates in the macro `wrap_pyfunction` (in Nightly builds, run with -Z macro-backtrace for more info)\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope","code":{"code":"E0599","explanation":"This error occurs when a method is used on a type which doesn't implement it:\n\nErroneous code example:\n\n```compile_fail,E0599\nstruct Mouth;\n\nlet x = Mouth;\nx.chocolate(); // error: no method named `chocolate` found for type `Mouth`\n               //        in the current scope\n```\n\nIn this case, you need to implement the `chocolate` method to fix the error:\n\n```\nstruct Mouth;\n\nimpl Mouth {\n    fn chocolate(&self) { // We implement the `chocolate` method here.\n        println!(\"Hmmm! I love chocolate!\");\n    }\n}\n\nlet x = Mouth;\nx.chocolate(); // ok!\n```\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":3773,"byte_end":3785,"line_start":113,"line_end":113,"column_start":7,"column_end":19,"is_primary":true,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_remove, m)?)?;","highlight_start":7,"highlight_end":19}],"label":"method not found in `&PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0599]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:113:7\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m113\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_remove, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m       \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mmethod not found in `&PyModule`\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied","code":{"code":"E0277","explanation":"You tried to use a type which doesn't implement some trait in a place which\nexpected that trait.\n\nErroneous code example:\n\n```compile_fail,E0277\n// here we declare the Foo trait with a bar method\ntrait Foo {\n    fn bar(&self);\n}\n\n// we now declare a function which takes an object implementing the Foo trait\nfn some_func<T: Foo>(foo: T) {\n    foo.bar();\n}\n\nfn main() {\n    // we now call the method with the i32 type, which doesn't implement\n    // the Foo trait\n    some_func(5i32); // error: the trait bound `i32 : Foo` is not satisfied\n}\n```\n\nIn order to fix this error, verify that the type you're using does implement\nthe trait. Example:\n\n```\ntrait Foo {\n    fn bar(&self);\n}\n\n// we implement the trait on the i32 type\nimpl Foo for i32 {\n    fn bar(&self) {}\n}\n\nfn some_func<T: Foo>(foo: T) {\n    foo.bar(); // we can now use this method since i32 implements the\n               // Foo trait\n}\n\nfn main() {\n    some_func(5i32); // ok!\n}\n```\n\nOr in a generic context, an erroneous code example would look like:\n\n```compile_fail,E0277\nfn some_func<T>(foo: T) {\n    println!(\"{:?}\", foo); // error: the trait `core::fmt::Debug` is not\n                           //        implemented for the type `T`\n}\n\nfn main() {\n    // We now call the method with the i32 type,\n    // which *does* implement the Debug trait.\n    some_func(5i32);\n}\n```\n\nNote that the error here is in the definition of the generic function. Although\nwe only call it with a parameter that does implement `Debug`, the compiler\nstill rejects the function. It must work with all possible input types. In\norder to make this example compile, we need to restrict the generic type we're\naccepting:\n\n```\nuse std::fmt;\n\n// Restrict the input type to types that implement Debug.\nfn some_func<T: fmt::Debug>(foo: T) {\n    println!(\"{:?}\", foo);\n}\n\nfn main() {\n    // Calling the method is still fine, as i32 implements Debug.\n    some_func(5i32);\n\n    // This would fail to compile now:\n    // struct WithoutDebug;\n    // some_func(WithoutDebug);\n}\n```\n\nRust only looks at the signature of the called function, as such it must\nalready specify all requirements that will be used for every type parameter.\n"},"level":"error","spans":[{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5200,"byte_end":5212,"line_start":153,"line_end":153,"column_start":13,"column_end":25,"is_primary":true,"text":[{"text":"            py_or_module,","highlight_start":13,"highlight_end":25}],"label":"the trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3786,"byte_end":3823,"line_start":113,"line_end":113,"column_start":20,"column_end":57,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_remove, m)?)?;","highlight_start":20,"highlight_end":57}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}},{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5125,"byte_end":5186,"line_start":152,"line_end":152,"column_start":9,"column_end":70,"is_primary":false,"text":[{"text":"        $crate::impl_::pyfunction::WrapPyFunctionArg::wrap_pyfunction(","highlight_start":9,"highlight_end":70}],"label":"required by a bound introduced by this call","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3786,"byte_end":3823,"line_start":113,"line_end":113,"column_start":20,"column_end":57,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_remove, m)?)?;","highlight_start":20,"highlight_end":57}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}}],"children":[{"message":"the following other types implement trait `WrapPyFunctionArg<'py, T>`:\n  &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  &pyo3::Bound<'py, pyo3::types::PyModule>\n  OnlyBound<T>\n  pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  pyo3::Bound<'py, pyo3::types::PyModule>\n  pyo3::Python<'py>","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0277]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:113:20\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m113\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_remove, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mthe trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mrequired by a bound introduced by this call\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mhelp\u001b[0m\u001b[0m: the following other types implement trait `WrapPyFunctionArg<'py, T>`:\u001b[0m\n\u001b[0m              &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              &pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              OnlyBound<T>\u001b[0m\n\u001b[0m              pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Python<'py>\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: this error originates in the macro `wrap_pyfunction` (in Nightly builds, run with -Z macro-backtrace for more info)\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope","code":{"code":"E0599","explanation":"This error occurs when a method is used on a type which doesn't implement it:\n\nErroneous code example:\n\n```compile_fail,E0599\nstruct Mouth;\n\nlet x = Mouth;\nx.chocolate(); // error: no method named `chocolate` found for type `Mouth`\n               //        in the current scope\n```\n\nIn this case, you need to implement the `chocolate` method to fix the error:\n\n```\nstruct Mouth;\n\nimpl Mouth {\n    fn chocolate(&self) { // We implement the `chocolate` method here.\n        println!(\"Hmmm! I love chocolate!\");\n    }\n}\n\nlet x = Mouth;\nx.chocolate(); // ok!\n```\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":3835,"byte_end":3847,"line_start":114,"line_end":114,"column_start":7,"column_end":19,"is_primary":true,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_exists, m)?)?;","highlight_start":7,"highlight_end":19}],"label":"method not found in `&PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0599]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:114:7\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m114\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_exists, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m       \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mmethod not found in `&PyModule`\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied","code":{"code":"E0277","explanation":"You tried to use a type which doesn't implement some trait in a place which\nexpected that trait.\n\nErroneous code example:\n\n```compile_fail,E0277\n// here we declare the Foo trait with a bar method\ntrait Foo {\n    fn bar(&self);\n}\n\n// we now declare a function which takes an object implementing the Foo trait\nfn some_func<T: Foo>(foo: T) {\n    foo.bar();\n}\n\nfn main() {\n    // we now call the method with the i32 type, which doesn't implement\n    // the Foo trait\n    some_func(5i32); // error: the trait bound `i32 : Foo` is not satisfied\n}\n```\n\nIn order to fix this error, verify that the type you're using does implement\nthe trait. Example:\n\n```\ntrait Foo {\n    fn bar(&self);\n}\n\n// we implement the trait on the i32 type\nimpl Foo for i32 {\n    fn bar(&self) {}\n}\n\nfn some_func<T: Foo>(foo: T) {\n    foo.bar(); // we can now use this method since i32 implements the\n               // Foo trait\n}\n\nfn main() {\n    some_func(5i32); // ok!\n}\n```\n\nOr in a generic context, an erroneous code example would look like:\n\n```compile_fail,E0277\nfn some_func<T>(foo: T) {\n    println!(\"{:?}\", foo); // error: the trait `core::fmt::Debug` is not\n                           //        implemented for the type `T`\n}\n\nfn main() {\n    // We now call the method with the i32 type,\n    // which *does* implement the Debug trait.\n    some_func(5i32);\n}\n```\n\nNote that the error here is in the definition of the generic function. Although\nwe only call it with a parameter that does implement `Debug`, the compiler\nstill rejects the function. It must work with all possible input types. In\norder to make this example compile, we need to restrict the generic type we're\naccepting:\n\n```\nuse std::fmt;\n\n// Restrict the input type to types that implement Debug.\nfn some_func<T: fmt::Debug>(foo: T) {\n    println!(\"{:?}\", foo);\n}\n\nfn main() {\n    // Calling the method is still fine, as i32 implements Debug.\n    some_func(5i32);\n\n    // This would fail to compile now:\n    // struct WithoutDebug;\n    // some_func(WithoutDebug);\n}\n```\n\nRust only looks at the signature of the called function, as such it must\nalready specify all requirements that will be used for every type parameter.\n"},"level":"error","spans":[{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5200,"byte_end":5212,"line_start":153,"line_end":153,"column_start":13,"column_end":25,"is_primary":true,"text":[{"text":"            py_or_module,","highlight_start":13,"highlight_end":25}],"label":"the trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3848,"byte_end":3885,"line_start":114,"line_end":114,"column_start":20,"column_end":57,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_exists, m)?)?;","highlight_start":20,"highlight_end":57}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}},{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5125,"byte_end":5186,"line_start":152,"line_end":152,"column_start":9,"column_end":70,"is_primary":false,"text":[{"text":"        $crate::impl_::pyfunction::WrapPyFunctionArg::wrap_pyfunction(","highlight_start":9,"highlight_end":70}],"label":"required by a bound introduced by this call","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3848,"byte_end":3885,"line_start":114,"line_end":114,"column_start":20,"column_end":57,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_exists, m)?)?;","highlight_start":20,"highlight_end":57}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}}],"children":[{"message":"the following other types implement trait `WrapPyFunctionArg<'py, T>`:\n  &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  &pyo3::Bound<'py, pyo3::types::PyModule>\n  OnlyBound<T>\n  pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  pyo3::Bound<'py, pyo3::types::PyModule>\n  pyo3::Python<'py>","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0277]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:114:20\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m114\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_exists, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mthe trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mrequired by a bound introduced by this call\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mhelp\u001b[0m\u001b[0m: the following other types implement trait `WrapPyFunctionArg<'py, T>`:\u001b[0m\n\u001b[0m              &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              &pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              OnlyBound<T>\u001b[0m\n\u001b[0m              pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Python<'py>\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: this error originates in the macro `wrap_pyfunction` (in Nightly builds, run with -Z macro-backtrace for more info)\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope","code":{"code":"E0599","explanation":"This error occurs when a method is used on a type which doesn't implement it:\n\nErroneous code example:\n\n```compile_fail,E0599\nstruct Mouth;\n\nlet x = Mouth;\nx.chocolate(); // error: no method named `chocolate` found for type `Mouth`\n               //        in the current scope\n```\n\nIn this case, you need to implement the `chocolate` method to fix the error:\n\n```\nstruct Mouth;\n\nimpl Mouth {\n    fn chocolate(&self) { // We implement the `chocolate` method here.\n        println!(\"Hmmm! I love chocolate!\");\n    }\n}\n\nlet x = Mouth;\nx.chocolate(); // ok!\n```\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":3897,"byte_end":3909,"line_start":115,"line_end":115,"column_start":7,"column_end":19,"is_primary":true,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_clone, m)?)?;","highlight_start":7,"highlight_end":19}],"label":"method not found in `&PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0599]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:115:7\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m115\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_clone, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m       \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mmethod not found in `&PyModule`\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied","code":{"code":"E0277","explanation":"You tried to use a type which doesn't implement some trait in a place which\nexpected that trait.\n\nErroneous code example:\n\n```compile_fail,E0277\n// here we declare the Foo trait with a bar method\ntrait Foo {\n    fn bar(&self);\n}\n\n// we now declare a function which takes an object implementing the Foo trait\nfn some_func<T: Foo>(foo: T) {\n    foo.bar();\n}\n\nfn main() {\n    // we now call the method with the i32 type, which doesn't implement\n    // the Foo trait\n    some_func(5i32); // error: the trait bound `i32 : Foo` is not satisfied\n}\n```\n\nIn order to fix this error, verify that the type you're using does implement\nthe trait. Example:\n\n```\ntrait Foo {\n    fn bar(&self);\n}\n\n// we implement the trait on the i32 type\nimpl Foo for i32 {\n    fn bar(&self) {}\n}\n\nfn some_func<T: Foo>(foo: T) {\n    foo.bar(); // we can now use this method since i32 implements the\n               // Foo trait\n}\n\nfn main() {\n    some_func(5i32); // ok!\n}\n```\n\nOr in a generic context, an erroneous code example would look like:\n\n```compile_fail,E0277\nfn some_func<T>(foo: T) {\n    println!(\"{:?}\", foo); // error: the trait `core::fmt::Debug` is not\n                           //        implemented for the type `T`\n}\n\nfn main() {\n    // We now call the method with the i32 type,\n    // which *does* implement the Debug trait.\n    some_func(5i32);\n}\n```\n\nNote that the error here is in the definition of the generic function. Although\nwe only call it with a parameter that does implement `Debug`, the compiler\nstill rejects the function. It must work with all possible input types. In\norder to make this example compile, we need to restrict the generic type we're\naccepting:\n\n```\nuse std::fmt;\n\n// Restrict the input type to types that implement Debug.\nfn some_func<T: fmt::Debug>(foo: T) {\n    println!(\"{:?}\", foo);\n}\n\nfn main() {\n    // Calling the method is still fine, as i32 implements Debug.\n    some_func(5i32);\n\n    // This would fail to compile now:\n    // struct WithoutDebug;\n    // some_func(WithoutDebug);\n}\n```\n\nRust only looks at the signature of the called function, as such it must\nalready specify all requirements that will be used for every type parameter.\n"},"level":"error","spans":[{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5200,"byte_end":5212,"line_start":153,"line_end":153,"column_start":13,"column_end":25,"is_primary":true,"text":[{"text":"            py_or_module,","highlight_start":13,"highlight_end":25}],"label":"the trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3910,"byte_end":3946,"line_start":115,"line_end":115,"column_start":20,"column_end":56,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_clone, m)?)?;","highlight_start":20,"highlight_end":56}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}},{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5125,"byte_end":5186,"line_start":152,"line_end":152,"column_start":9,"column_end":70,"is_primary":false,"text":[{"text":"        $crate::impl_::pyfunction::WrapPyFunctionArg::wrap_pyfunction(","highlight_start":9,"highlight_end":70}],"label":"required by a bound introduced by this call","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3910,"byte_end":3946,"line_start":115,"line_end":115,"column_start":20,"column_end":56,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_clone, m)?)?;","highlight_start":20,"highlight_end":56}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}}],"children":[{"message":"the following other types implement trait `WrapPyFunctionArg<'py, T>`:\n  &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  &pyo3::Bound<'py, pyo3::types::PyModule>\n  OnlyBound<T>\n  pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  pyo3::Bound<'py, pyo3::types::PyModule>\n  pyo3::Python<'py>","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0277]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:115:20\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m115\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_clone, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mthe trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mrequired by a bound introduced by this call\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mhelp\u001b[0m\u001b[0m: the following other types implement trait `WrapPyFunctionArg<'py, T>`:\u001b[0m\n\u001b[0m              &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              &pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              OnlyBound<T>\u001b[0m\n\u001b[0m              pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Python<'py>\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: this error originates in the macro `wrap_pyfunction` (in Nightly builds, run with -Z macro-backtrace for more info)\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope","code":{"code":"E0599","explanation":"This error occurs when a method is used on a type which doesn't implement it:\n\nErroneous code example:\n\n```compile_fail,E0599\nstruct Mouth;\n\nlet x = Mouth;\nx.chocolate(); // error: no method named `chocolate` found for type `Mouth`\n               //        in the current scope\n```\n\nIn this case, you need to implement the `chocolate` method to fix the error:\n\n```\nstruct Mouth;\n\nimpl Mouth {\n    fn chocolate(&self) { // We implement the `chocolate` method here.\n        println!(\"Hmmm! I love chocolate!\");\n    }\n}\n\nlet x = Mouth;\nx.chocolate(); // ok!\n```\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":3958,"byte_end":3970,"line_start":116,"line_end":116,"column_start":7,"column_end":19,"is_primary":true,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_resize, m)?)?;","highlight_start":7,"highlight_end":19}],"label":"method not found in `&PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0599]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: no method named `add_function` found for reference `&pyo3::types::PyModule` in the current scope\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:116:7\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m116\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_resize, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m       \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mmethod not found in `&PyModule`\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied","code":{"code":"E0277","explanation":"You tried to use a type which doesn't implement some trait in a place which\nexpected that trait.\n\nErroneous code example:\n\n```compile_fail,E0277\n// here we declare the Foo trait with a bar method\ntrait Foo {\n    fn bar(&self);\n}\n\n// we now declare a function which takes an object implementing the Foo trait\nfn some_func<T: Foo>(foo: T) {\n    foo.bar();\n}\n\nfn main() {\n    // we now call the method with the i32 type, which doesn't implement\n    // the Foo trait\n    some_func(5i32); // error: the trait bound `i32 : Foo` is not satisfied\n}\n```\n\nIn order to fix this error, verify that the type you're using does implement\nthe trait. Example:\n\n```\ntrait Foo {\n    fn bar(&self);\n}\n\n// we implement the trait on the i32 type\nimpl Foo for i32 {\n    fn bar(&self) {}\n}\n\nfn some_func<T: Foo>(foo: T) {\n    foo.bar(); // we can now use this method since i32 implements the\n               // Foo trait\n}\n\nfn main() {\n    some_func(5i32); // ok!\n}\n```\n\nOr in a generic context, an erroneous code example would look like:\n\n```compile_fail,E0277\nfn some_func<T>(foo: T) {\n    println!(\"{:?}\", foo); // error: the trait `core::fmt::Debug` is not\n                           //        implemented for the type `T`\n}\n\nfn main() {\n    // We now call the method with the i32 type,\n    // which *does* implement the Debug trait.\n    some_func(5i32);\n}\n```\n\nNote that the error here is in the definition of the generic function. Although\nwe only call it with a parameter that does implement `Debug`, the compiler\nstill rejects the function. It must work with all possible input types. In\norder to make this example compile, we need to restrict the generic type we're\naccepting:\n\n```\nuse std::fmt;\n\n// Restrict the input type to types that implement Debug.\nfn some_func<T: fmt::Debug>(foo: T) {\n    println!(\"{:?}\", foo);\n}\n\nfn main() {\n    // Calling the method is still fine, as i32 implements Debug.\n    some_func(5i32);\n\n    // This would fail to compile now:\n    // struct WithoutDebug;\n    // some_func(WithoutDebug);\n}\n```\n\nRust only looks at the signature of the called function, as such it must\nalready specify all requirements that will be used for every type parameter.\n"},"level":"error","spans":[{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5200,"byte_end":5212,"line_start":153,"line_end":153,"column_start":13,"column_end":25,"is_primary":true,"text":[{"text":"            py_or_module,","highlight_start":13,"highlight_end":25}],"label":"the trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3971,"byte_end":4008,"line_start":116,"line_end":116,"column_start":20,"column_end":57,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_resize, m)?)?;","highlight_start":20,"highlight_end":57}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}},{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":5125,"byte_end":5186,"line_start":152,"line_end":152,"column_start":9,"column_end":70,"is_primary":false,"text":[{"text":"        $crate::impl_::pyfunction::WrapPyFunctionArg::wrap_pyfunction(","highlight_start":9,"highlight_end":70}],"label":"required by a bound introduced by this call","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3971,"byte_end":4008,"line_start":116,"line_end":116,"column_start":20,"column_end":57,"is_primary":false,"text":[{"text":"    m.add_function(wrap_pyfunction!(benchmark_resize, m)?)?;","highlight_start":20,"highlight_end":57}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"wrap_pyfunction!","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-0.22.5\\src\\macros.rs","byte_start":4485,"byte_end":4513,"line_start":136,"line_end":136,"column_start":1,"column_end":29,"is_primary":false,"text":[{"text":"macro_rules! wrap_pyfunction {","highlight_start":1,"highlight_end":29}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}}],"children":[{"message":"the following other types implement trait `WrapPyFunctionArg<'py, T>`:\n  &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  &pyo3::Bound<'py, pyo3::types::PyModule>\n  OnlyBound<T>\n  pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\n  pyo3::Bound<'py, pyo3::types::PyModule>\n  pyo3::Python<'py>","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0277]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: the trait bound `&pyo3::types::PyModule: WrapPyFunctionArg<'_, _>` is not satisfied\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:116:20\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m116\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m    m.add_function(wrap_pyfunction!(benchmark_resize, m)?)?;\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mthe trait `WrapPyFunctionArg<'_, _>` is not implemented for `&pyo3::types::PyModule`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m                    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mrequired by a bound introduced by this call\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mhelp\u001b[0m\u001b[0m: the following other types implement trait `WrapPyFunctionArg<'py, T>`:\u001b[0m\n\u001b[0m              &pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              &pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              OnlyBound<T>\u001b[0m\n\u001b[0m              pyo3::Borrowed<'_, 'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Bound<'py, pyo3::types::PyModule>\u001b[0m\n\u001b[0m              pyo3::Python<'py>\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: this error originates in the macro `wrap_pyfunction` (in Nightly builds, run with -Z macro-backtrace for more info)\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"the trait bound `&pyo3::types::PyModule: From<BoundRef<'_, '_, pyo3::types::PyModule>>` is not satisfied","code":{"code":"E0277","explanation":"You tried to use a type which doesn't implement some trait in a place which\nexpected that trait.\n\nErroneous code example:\n\n```compile_fail,E0277\n// here we declare the Foo trait with a bar method\ntrait Foo {\n    fn bar(&self);\n}\n\n// we now declare a function which takes an object implementing the Foo trait\nfn some_func<T: Foo>(foo: T) {\n    foo.bar();\n}\n\nfn main() {\n    // we now call the method with the i32 type, which doesn't implement\n    // the Foo trait\n    some_func(5i32); // error: the trait bound `i32 : Foo` is not satisfied\n}\n```\n\nIn order to fix this error, verify that the type you're using does implement\nthe trait. Example:\n\n```\ntrait Foo {\n    fn bar(&self);\n}\n\n// we implement the trait on the i32 type\nimpl Foo for i32 {\n    fn bar(&self) {}\n}\n\nfn some_func<T: Foo>(foo: T) {\n    foo.bar(); // we can now use this method since i32 implements the\n               // Foo trait\n}\n\nfn main() {\n    some_func(5i32); // ok!\n}\n```\n\nOr in a generic context, an erroneous code example would look like:\n\n```compile_fail,E0277\nfn some_func<T>(foo: T) {\n    println!(\"{:?}\", foo); // error: the trait `core::fmt::Debug` is not\n                           //        implemented for the type `T`\n}\n\nfn main() {\n    // We now call the method with the i32 type,\n    // which *does* implement the Debug trait.\n    some_func(5i32);\n}\n```\n\nNote that the error here is in the definition of the generic function. Although\nwe only call it with a parameter that does implement `Debug`, the compiler\nstill rejects the function. It must work with all possible input types. In\norder to make this example compile, we need to restrict the generic type we're\naccepting:\n\n```\nuse std::fmt;\n\n// Restrict the input type to types that implement Debug.\nfn some_func<T: fmt::Debug>(foo: T) {\n    println!(\"{:?}\", foo);\n}\n\nfn main() {\n    // Calling the method is still fine, as i32 implements Debug.\n    some_func(5i32);\n\n    // This would fail to compile now:\n    // struct WithoutDebug;\n    // some_func(WithoutDebug);\n}\n```\n\nRust only looks at the signature of the called function, as such it must\nalready specify all requirements that will be used for every type parameter.\n"},"level":"error","spans":[{"file_name":"src\\pylib.rs","byte_start":3574,"byte_end":3585,"line_start":109,"line_end":109,"column_start":1,"column_end":12,"is_primary":true,"text":[{"text":"#[pymodule]","highlight_start":1,"highlight_end":12}],"label":"the trait `From<BoundRef<'_, '_, pyo3::types::PyModule>>` is not implemented for `&pyo3::types::PyModule`, which is required by `BoundRef<'_, '_, pyo3::types::PyModule>: Into<_>`","suggested_replacement":null,"suggestion_applicability":null,"expansion":{"span":{"file_name":"src\\pylib.rs","byte_start":3574,"byte_end":3585,"line_start":109,"line_end":109,"column_start":1,"column_end":12,"is_primary":false,"text":[{"text":"#[pymodule]","highlight_start":1,"highlight_end":12}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null},"macro_decl_name":"#[pymodule]","def_site_span":{"file_name":"C:\\Users\\cleme\\.cargo\\registry\\src\\index.crates.io-6f17d22bba15001f\\pyo3-macros-0.22.5\\src/lib.rs","byte_start":2030,"byte_end":2099,"line_start":40,"line_end":40,"column_start":1,"column_end":70,"is_primary":false,"text":[{"text":"pub fn pymodule(args: TokenStream, input: TokenStream) -> TokenStream {","highlight_start":1,"highlight_end":70}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}}}],"children":[{"message":"required for `BoundRef<'_, '_, pyo3::types::PyModule>` to implement `Into<&pyo3::types::PyModule>`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror[E0277]\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: the trait bound `&pyo3::types::PyModule: From<BoundRef<'_, '_, pyo3::types::PyModule>>` is not satisfied\u001b[0m\n\u001b[0m   \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m--> \u001b[0m\u001b[0msrc\\pylib.rs:109:1\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m\u001b[1m\u001b[38;5;14m109\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m#[pymodule]\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9m^^^^^^^^^^^\u001b[0m\u001b[0m \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;9mthe trait `From<BoundRef<'_, '_, pyo3::types::PyModule>>` is not implemented for `&pyo3::types::PyModule`, which is required by `BoundRef<'_, '_, pyo3::types::PyModule>: Into<_>`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m|\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: required for `BoundRef<'_, '_, pyo3::types::PyModule>` to implement `Into<&pyo3::types::PyModule>`\u001b[0m\n\u001b[0m    \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;14m= \u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15mnote\u001b[0m\u001b[0m: this error originates in the attribute macro `pymodule` (in Nightly builds, run with -Z macro-backtrace for more info)\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"aborting due to 14 previous errors; 1 warning emitted","code":null,"level":"error","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;9merror\u001b[0m\u001b[0m\u001b[1m\u001b[38;5;15m: aborting due to 14 previous errors; 1 warning emitted\u001b[0m\n\n"}
{"$message_type":"diagnostic","message":"Some errors have detailed explanations: E0277, E0432, E0599.","code":null,"level":"failure-note","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;15mSome errors have detailed explanations: E0277, E0432, E0599.\u001b[0m\n"}
{"$message_type":"diagnostic","message":"For more information about an error, try `rustc --explain E0277`.","code":null,"level":"failure-note","spans":[],"children":[],"rendered":"\u001b[0m\u001b[1m\u001b[38;5;15mFor more information about an error, try `rustc --explain E0277`.\u001b[0m\n"}
